Non-latin text-recognition support

spetrey opened the pull request:
Closes #27

## What does this PR do?

This PR adds official text-recognition support for:
```
GoogleMLKit/TextRecognitionChinese
GoogleMLKit/TextRecognitionDevanagari
GoogleMLKit/TextRecognitionJapanese
GoogleMLKit/TextRecognitionKorean
```

## How to use

```tsx
const result = await TextRecognition.recognize(imagePath) // Latin text-recognition
...
const result = await TextRecognition.recognize(imagePath, 'Chinese') // Chinese text-recognition
...
const result = await TextRecognition.recognize(imagePath, 'Devanagari') // Devanagari text-recognition
...
const result = await TextRecognition.recognize(imagePath, 'Japanese') // Japanese text-recognition
...
const result = await TextRecognition.recognize(imagePath, 'Korean') // Korean text-recognition
```

## Results:

**Japanese and Devanagari**
<img width="420" alt="Screenshot 2023-04-04 at 7 36 41 PM" src="https://user-images.githubusercontent.com/6411509/229947177-293c5da3-97b2-417e-b54b-b058ebfe049d.png">

**Chinese**
<img width="420" alt="Screenshot 2023-04-04 at 7 34 38 PM" src="https://user-images.githubusercontent.com/6411509/229947180-b83a0bb6-c6a7-4179-8833-9798263c7d98.png">

**Korean**
<img width="420" alt="Screenshot 2023-04-04 at 7 33 41 PM" src="https://user-images.githubusercontent.com/6411509/229947184-a1cb2716-7762-4f2b-8f25-664809dcb75f.png">


This is my first PR here! I will be the first to admit that Objective-C is not my strongest language, so I welcome any and all feedback! I would love for someone to pull this branch down and take it for a spin. My initial testing yielded some good results but would love someone to double-check my work and give me a sanity check 😄

a7medev commented on text-recognition/index.d.ts:
```diff
@@ -62,9 +62,17 @@ interface ITextRecognition {
    *
    * @param imageURL The URL/Path of the image to process
+   *
+   * @param language @optional The language to use for non-latin
+   * text recognition. The default is English. Available languages:
+   * - Chinese
+   * - Devanagari
+   * - Japanese
+   * - Korean
+   *
    * @returns Text recognition result
    */
-  recognize: (imageURL: string) => Promise<TextRecognitionResult>;
+  recognize: (imageURL: string, language?: string | null) => Promise<TextRecognitionResult>;
```
I'd say let's make `language` an enum instead. See [@react-native-ml-kit/translate-text's `TranslateLanguage` enum](https://github.com/a7medev/react-native-ml-kit/blob/62ad4f3db2b083b8f527c36212ea7cc65a410cb8/translate-text/index.js#L9).

spetrey replied:
Totally onboard with this change. Will address tonight 👍

a7medev commented on text-recognition/RNMLKitTextRecognition.podspec:
```diff
@@ -22,14 +22,14 @@ Pod::Spec.new do |s|
   # To recognize Chinese script
-  s.dependency 'GoogleMLKit/TextRecognitionChinese', '2.6.0'
+  s.dependency 'GoogleMLKit/TextRecognitionChinese', '4.0.0'
   # To recognize Devanagari script
-  s.dependency 'GoogleMLKit/TextRecognitionDevanagari', '2.6.0'
+  s.dependency 'GoogleMLKit/TextRecognitionDevanagari', '4.0.0'
   # To recognize Japanese script
-  s.dependency 'GoogleMLKit/TextRecognitionJapanese', '2.6.0'
+  s.dependency 'GoogleMLKit/TextRecognitionJapanese', '4.0.0'
   # To recognize Korean script
-  s.dependency 'GoogleMLKit/TextRecognitionKorean', '2.6.0'
+  s.dependency 'GoogleMLKit/TextRecognitionKorean', '4.0.0'
```
Let's extract the dependency updates part into another PR updating dependencies for all the packages as they need to be kept in-sync to not break apps that use multiple packages.

spetrey commented:
@a7medev great feedback on the PR! Much appreciated. I addressed your comments and went a little further to do some additional cleanup. I also figured I would rename the `language` param to `script` [here](https://github.com/a7medev/react-native-ml-kit/pull/28/files#diff-e27ca47e4e001b5ab59852261c0b71b0a1b90eaba5f56c7a136edc762562678fR85) because it's based on Script names instead of Language names. See this table from ML Kit's text recognition v2 docs:

https://developers.google.com/ml-kit/vision/text-recognition/v2/languages#supported-langs

Let me know if you see anything else that needs addressing :)

a7medev commented on text-recognition/index.d.ts:
```diff
@@ -57,14 +57,33 @@ export interface TextRecognitionResult {
   blocks: TextBlock[];
 }
 
+export enum TranslateRecognitionScript {
```
I think it should be `TextRecognitionScript`.

```suggestion
export enum TextRecognitionScript {
```

a7medev commented:
@spetrey I think will also need to support for Android as well to keep both platforms consistent. What do you think?

spetrey commented:
@a7medev apologies on the delay getting back to you. I do agree that supporting Android is important. Unfortunately, I have never worked on any Java projects before this, so I may be unable to contribute to supporting Android on this PR (in the short-term) 😞 

What're your thoughts on merging this PR, then creating an Issue to track adding support for `TextRecognition` on Android at a later date?

a7medev commented:
@spetrey Sounds good. Thanks for your contribution! 🙏
I'll test this PR on iOS and merge it and will work on another PR for Android.

spetrey commented:
Thank you for taking the time to review! Let me know if you encounter any issues testing @a7medev, happy to resolve anything that comes up 👍

a7medev commented on text-recognition/ios/TextRecognition.m:
```diff
@@ -75,15 +79,35 @@ - (NSDictionary*)blockToDict: (MLKTextBlock*)block {
+    MLKTextRecognizerOptions *options = nil;
+
+    // if the language param isn't specified, we can assume the user requirement is Latin text recognition
+    if (script == nil) {
```
```suggestion
    if (script == nil || [script isEqualToString:@"Latin"]) {
```

spetrey replied:
Ah, good suggestion!

a7medev reviewed (changes requested):
Calling `TextRecognition.recognize(imagePath)` without the `script` parameter doesn't work and throws this error: `TextRecognition.recognize was called with 1 arguments but expected 2 arguments.`

We need to make sure this feature is backward-compatible by providing the default value of `"Latin"` on iOS and not passing the `script` parameter on Android which will require us to make a wrapper method around the `TextRecognition.recognize` native method but without changing the API to not break the current version of the library. I'll handle that case and merge the PR once ready.

Also, there's another change for making the `"Latin"` enum value work, please check it.

Thanks so much for your contribution! 🙏🏼

spetrey commented:
Great feedback @a7medev! Thanks for double-checking my contribution and ensuring backward compat on Android 🙏

a7medev reviewed (approved):
Looks great, let's ship it in the next release! 🚀

cardanoeconews commented:
Hi @a7medev and @spetrey , Just checking in to see if Android support for non-latin has been added or if you've started on that MR @a7medev ? It's something needed before I can ship my product as there are a ton of android users in South Korea. Would love any help, happy to sponsor!

a7medev commented:
Hi @cardanoeconews! Unfortunately, I haven't had much time to work on adding Android support lately, will work on it once I have some time. 🙏🏼 Happy to know you are using it in production!

I don't have GitHub Sponsors set up at the moment, but I'm working on it—having some support would actually allow me to spend more time working on these projects and delivering more features.

a7medev commented:
Wow, that was fast! 😅 I've set up GitHub Sponsors, you can sponsor my work on React Native ML Kit and other packages through [my Sponsors page](https://github.com/sponsors/a7medev)!

a7medev commented:
Quick update! Non-Latin script Android support just landed in `@react-native-ml-kit/text-recognition` v1.5.0 🚀
